Implement RedisCacheService

dlloyd1-ch opened the pull request:
Implementation of Redis Cache Service
Added config
Added integration test using Redis test Container

dlloyd1-ch committed:
Initial draft of RedisCacheService

dlloyd1-ch committed:
Change file names to same as pacakage name

dlloyd1-ch committed:
Added time to live entry to Redis Cache Service

Change Read function to use ZRANGEBYSCORE

dlloyd1-ch committed:
Implemented delete functionality for expired entries

dlloyd1-ch committed:
Added config and code tidy up.

dlloyd1-ch committed:
Added check for empty array in Delete function

dlgroves-ch commented on cache/cache.go:
```diff
@@ -0,0 +1,87 @@
+	}
+}
+
+func (r RedisCacheService) Create(key string, delta string, score int64) error {
```
Am a little concerned this approach may cause duplicate offsets to be added to the second sorted set - is it feasible to replace this with an ADD + EXPIRE?

dlloyd1-ch replied:
EXPIRE works at key level. Currently all deltas for a stream are stored in the same key so using EXPIRE would wipe all the data at expiry time.. We could create a new key for each delta using the offset as part of the key but this would mean we would lose the ability to use the RANGE command to retrieve the data.

dlloyd1-ch replied:
Refactored so that each delta is stored in as a Redis string value with it's own key.
An expiry time is set on the key which Redis automatically handles so no need for manual delete.
The offsets are stored in a sorted set with the offset as key and the corresponding delta key at its value.
The Read() method:
1) Retrieves a list of keys for the deltas from the sorted set based on the given offset
2) Reads and returns the deltas for each delta key

dlloyd1-ch committed:
Implemented redis testcontainer for integration tests

Ugraded to go version 1.15 (mimimum required is 1.13)

dlloyd1-ch committed:
Added test for duplicate deltas

General code tidy up.

dlloyd1-ch committed:
Refacored RedisCache service to use Strings and expire functionality

Each delta is stored in as a Redis string value with it's own key.
An expiry time is set on the key which Redis automatically handles so no need for manual delete.
The offsets are stored in a stoerted set with the oddset as key and the corresponding delta key at its value.
